File: homework/thinhbka/lesson-6-binary-search-bst/1214. Two Sum BSTs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def twoSumBSTs(self, root1: Optional[TreeNode], root2: Optional[TreeNode], target: int) -> bool:
        values1 = {}
        def dfs(node):
            values1[node.val] = True
            if node.left:
                dfs(node.left)
            if node.right:
                dfs(node.right)
        dfs(root1)
        stack = [root2]
        while stack:
            node = stack.pop()
            if target- node.val in values1:
                return True
            if node.left:stack.append(node.left)
            if node.right:stack.append(node.right)
        return False

File: homework/Truong_Duong/Lesson_9/1214_Two_Sum_BSTs.js
// Solution for https://leetcode.com/problems/two-sum-bsts/

var twoSumBSTs = function(root1, root2, target) {
  const dict = {};

  const root1Traversal = (root) => {
    if(root == null) return;
    dict[root.val] = 1;
    root1Traversal(root.left);
    root1Traversal(root.right);
  }

  root1Traversal(root1);

  const root2Traversal = (root) => {
    if(root == null) return false;
    if((target - root.val) in dict) return true;

    if(root2Traversal(root.left)) return true;
    if(root2Traversal(root.right)) return true;
    return false;
  }

  return root2Traversal(root2);
};